### ISSUE TITLE
[Feature Requests] New Random Game Dialog

### ISSUE THREAD

**Issue by jasonkhanlar** [widelands/widelands#5522]
Single Player... -> New Random Game [Z]

1. Random number: Add a button to the right to randomly generate a new number when clicked.
    - https://github.com/widelands/widelands/issues/5525

2. Random number: When focus is on the textbox, allow pushing enter to submit the form to generate a new game.
    - https://github.com/widelands/widelands/issues/5526

3. Random number: Sometimes (not always, but not sure how to always reproduce) when fiddling around with the fields, and attempting to generate a random map results in Map Generation Error, clicking OK to close that dialog, then immediately clicking on the Random number text input field in the whitespace to the right of the numbers, pushing backspace one time (to get rid of the last number) does not get rid of the last number (again, sometimes, not always), but pushing backspace a second time then gets rid of the last number. Actually... instead of submitting the form each time, simply clicking on the checkbox for Island mode and then clicking on the Random number text field and pushing backspace, sometimes it works, other times the first keypress does nothing.
    - https://github.com/widelands/widelands/issues/5527

4. Randomly generated map:
    
     1.
        - Width: 512
        - Height: 512
        - Players: 2
        - Terrain Distribution:
            - Water: 0%
            - Land: 30%
            - Wasteland: 0%
            - Mountains: 70%
        - Island Mode: Off/No/Deselected
        - Random number: 16537704
        
        This map appears that player 2 is surrounded by mountains and unable to reach the opponent. I tried to expand as much as I could with Citadels that conquers 12 spaces. ![https://i.imgur.com/oQaFy7Z.png](https://i.imgur.com/oQaFy7Z.png)

    2.
        - 512x512, 2 players, 0% water, 25% land, 10% wasteland, 65% mountains, not an island, random number: 16537
        
        same issue, neither player can access the rest of the map

    4.
        - 512x512, 2 players, 60% water, 35% land, 5% wasteland, not an island, random number: 1051009590
        
        one player is trapped on tiny island

5. In the New Random Game configuration screen, remember last used settings.
    - https://github.com/widelands/widelands/issues/5528

6. Alternative map generators: For example see Minetest https://minetest.net/#features "Choose from many different fast map generators included in the engine, or install one as a mod." Also see https://wiki.minetest.net/Map_generator

    - https://github.com/widelands/widelands/issues/5529

7. Random number: When double clicking on the text input field, highlight all the text. Clicking and dragging selects. CTRL+A selects all. Double click should too.
    
    - Note that even when pushing CTRL+A or clicking and dragging to select all (or part of) the text, when typing a number expecting to replace/overwrite the value, instead it doesn't overwrite, and merely inserts the character where the cursor appears to be. However, when pushing the backspace or delete keys, apparently all of the selected text gets removed. This seems like the backspace/delete keys are treated differently than other keypresses.
    - Also suggestion: CTRL+Backspace should get rid of all the text to the left of the cursor, and CTRL+Delete all text to the right.

    ----
    - https://github.com/widelands/widelands/issues/5530

8. Random number: Prevent nonnumbers from showing up in the text input field
    - https://github.com/widelands/widelands/issues/5530#issuecomment-1223154684
9. Map ID: Disable editing this value
    - https://github.com/widelands/widelands/issues/5530#issuecomment-1223154684

**Comment by jasonkhanlar** [comment]
I just looked at the Editor... -> Random Map feature. After I generate a map, then select Main Menu -> New Random Map, same as # 5 in main issue, remember last used settings.

Also in 512x512 generated map with 100% mountain (not an island), saving and playing a map (even though no suitable place for the players, when playing this map and speeding up the game to 100x+, 

> WARNING:  2: AI: game speed too high, jobs are too late (now XX seconds)

but the AI is doing nothing, because it can't do anything.

Also I can't tell the different too easily about 100% wasteland versus 100% mountain.  Trying generating a few different combinations with wasteland/mountains to figure out the difference, I can't figure it out. I can generate 100% mountain, but I can't generate 100% wasteland, only 70%. I see some difference. For example, instead of Meadow 2, I see Barren Steppe, but not all the meadow, only some, definitely not 70%. It seems that wasteland is mostly mountains.

Climate: Summer, Resources: Low, 30% water, 70% wasteland, random number: 0 (top)

Climate: Summer, Resources: Low, 30% water, 70% mountain, random number: 0 (bottom)

![https://i.imgur.com/AVDoU7B.jpg](https://i.imgur.com/AVDoU7B.jpg)

Climate summer is selected, but I see snow.

----

Climate: Summer, Resources: Low, 30% land, 70% wasteland, random number: 0 (top)

Climate: Summer, Resources: Low, 30% land, 70% mountain, random number: 0 (bottom)

![https://i.imgur.com/WvbcceG.jpg](https://i.imgur.com/WvbcceG.jpg)

----

Comparing 100% land summer vs winter, I can see difference. I think this might be because land terrain adjusts for the season, but wasteland and mountain terrains do not. Let's find out:

Climate: Summer, Resources: Low, 100% mountain, random number: 0 (top)

Climate: Winter, Resources: Low, 100% mountain, random number: 0 (bottom)

![https://i.imgur.com/sDg2ehw.jpg](https://i.imgur.com/sDg2ehw.jpg) 

----

Also I just noticed the wasteland is a climate AND a terrain. That seems confusing without yet seeing what this means.

- See https://i.imgur.com/kr5Bq91.jpg for 100% land, climate: summer (top) vs wasteland (bottom)
- See below for 30% land, 70% wasteland
    - climate: summer
      ![https://i.imgur.com/HrjbhBD.png](https://i.imgur.com/HrjbhBD.png)
    - climate: wasteland (now [THIS](https://cdn.discordapp.com/attachments/349769190225346562/1011073609936142456/widelands_wasteland.gif) is AWESOME! click to see animation)
      ![http://i.imgur.com/hBgyyf4h.gif](http://i.imgur.com/hBgyyf4h.gif)

----

Aha! Climate: Wasteland and comparing 70%wasteland/30%mountain to 0%wasteland/100% mountain, they are the exact same thing! Lol what? Okay, then how about comparing the other climates?

- [Exact same!] Climate: Summer, 70%wasteland/30%mountain vs 0%wasteland/100% mountain
- [Exact same!] Climate: Winter, 70%wasteland/30%mountain vs 0%wasteland/100% mountain
- [Exact same!] Climate: Desert, 70%wasteland/30%mountain vs 0%wasteland/100% mountain

So then again I am confused (partially/mostly) as to what the wasteland terrain does differently (other than a tiny bit I noticed and described above, with screenshots, but even that tiny/minor difference does not equal the 70% as it should appear to be)

----

Editor... -> Random Map: Generate any map -> Main Menu -> Map Options -> Teams tab: The Add lineup button, I clicked it like 200 times and I was able to add ~200 blank lineups, filling out a few of them. I'm not sure how many maximum lineups would be technically reasonable to consider, but 200 seems like overkill. Perhaps limit the amount of lineups that can be created?

----

Editor... -> Random Map: Generate any map -> Tools -> Terrain: change order of tabs: Summer, Winter, Wasteland, Desert - to match the order that appears in New Random Map dialog Climate dropdown menu

**Comment by hessenfarmer** [comment]
Hi, 
well a lot of remarks to answer so I start with the easiest. 
> 
> > WARNING:  2: AI: game speed too high, jobs are too late (now XX seconds)
> 
> but the AI is doing nothing, because it can't do anything.
> 

In fact this warning is about the jobs of the AI. The jobs of the AI are calcualting different aspects of the game in scheduled regular intervals. Calculating these jobs needs prozessor time whis is real time. The jobs are scheduled in game time to make for good decisions when running the game at real time. If you accelerate the game time the jobs need to be calcualted more frequently in real time, up to the point where there isn't sufficient time available for calcualting this. 
So the jobs are running always even if the result is we can't do anything.

**Comment by frankystone** [comment]
In general the random map generator is a bit neglected… and probably buggy. See [bug reports](https://github.com/widelands/widelands/issues?q=is%3Aissue+is%3Aopen+"random+map")

Some history (you may exchange the term "world" with "climate"):  
Formerly widelands could only work with maps which are made out of one of the worlds (Summer, Wasteland, Winter and Desert). To make a new map one had to choose one of this worlds before editing a map and could only use terrains from this world. So some of the worlds have snow as terrain, e.g. summer. Today this restriction is gone and a map author can choose every terrain of every world in one map (and additionally terrains from add-ons). This is called "One World". One can see this on the [maps page](https://www.widelands.org/maps/) of our Homepage: New maps are tagged with term "One World" old maps are tagged with the world-name the map was made with, e.g. "Wasteland". The distinction between the worlds is still part of the widelands ui, reflected in the 4 (world) tabs of the terrain menu in the editor.

**Comment by hessenfarmer** [comment]
as this issue is in fact a mixture between some questions and some feature requests it is hard to track waht shall be / will be implemented and what will not be implemented. @jasonkhanlar can you split the clear feature request from the rest and open a new issue for each of them?

**Comment by jasonkhanlar** [comment]
> as this issue is in fact a mixture between some questions and some feature requests it is hard to track waht shall be / will be implemented and what will not be implemented. @jasonkhanlar can you split the clear feature request from the rest and open a new issue for each of them?

Sure! Not quite right away, but just moments before I got started with preparing this, I thought it would be easier to edit this single report with things I considered that may or may not be worthy of addressing (as I was learning/realizing along the way), rather than creating a bunch of individual issues. Mostly this is because I am still relatively new to tinkering with Widelands, despite having first discovered and played Widelands 10 years ago.

I initially concentrated specifically on the new random game aspect, but then I saw the editor, and there is a little overlap, so I posted a few separate things as a comment. Later I'll work on splitting apart a few single individual issues into their own posts, and then I'll edit here to link back to each so that things aren't too confusing.
